### PR TITLE
[RFC] Wind Waker: Shadow link

### PR DESCRIPTION
**Pull request by phire**
I think this totally improves Wind Waker with a new and unique look at the story of Shadow Link,

gzle01-12

Also Improved other titles, Mario Kart 64 takes on the look of an Ink wash painting. 

nabe01-2

This popped up during the development of UberShaders, was too good not to save for prosperity. Not for merging.

### CODE REVIEW

**Comment by phire** [comment]
Btw, this is mostly my WIP Swizzling support, with one change. @degasus is there any chance you can spot my bug?

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+		"			// TODO: there is an optional perspective divide here (not to mention all of indirect)\n"
+		"			int2 fixedPoint_uv = itrunc(tex[tex_coord].xy * " I_TEXDIMS"[sampler_num].zw * 128.0);\n"
+		"			float2 uv = (float2(fixedPoint_uv) / 128.0) * " I_TEXDIMS"[sampler_num].xy;\n"
+		"\n"
+		"			int4 color = sampleTexture(sampler_num, uv);\n"
+		"\n"
+		"			uint swap = %s;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.tswap).c_str());
+	out.Write(
+		//"			s.TexColor.r = Swap(0, color);\n");
+		"			s.TexColor.r = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2]", TevKSel().swap1).c_str());
```
Comment (phire) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: These should be Swap(%s, color) instead of swap(3u-%s, color)

The commented out swap are hard-coded for TexColor.rgba = color.rgba, which makes it behave the same as the previous commit.

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+		"			// TODO: there is an optional perspective divide here (not to mention all of indirect)\n"
+		"			int2 fixedPoint_uv = itrunc(tex[tex_coord].xy * " I_TEXDIMS"[sampler_num].zw * 128.0);\n"
+		"			float2 uv = (float2(fixedPoint_uv) / 128.0) * " I_TEXDIMS"[sampler_num].xy;\n"
+		"\n"
+		"			int4 color = sampleTexture(sampler_num, uv);\n"
+		"\n"
+		"			uint swap = %s;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.tswap).c_str());
+	out.Write(
+		//"			s.TexColor.r = Swap(0, color);\n");
+		"			s.TexColor.r = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2u]", TevKSel().swap1).c_str());
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Please don't hide abgr vs rgba convertion here in "3u-%s", just move it in Swap() itself

btw, is this needed at all? https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoCommon/PixelShaderGen.cpp#L498 shadergen also uses rgba

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kcsel1).c_str(),
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kasel1).c_str());
+	out.Write(
+		"\n");
+
+	out.Write(
+		"		// Select Ras for stage\n"
+		"		uint ras = %s;\n", BitfieldExtract("order", TwoTevStageOrders().colorchan0).c_str());
+	out.Write(
+		"		if (ras < 2u) { // Lighting Channel 0 or 1\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: What was wrong with the switch here?
```
switch (ras) {
case 0u:
case 1u:
  int4 color = ....
```

Comment (phire) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: I avoided that because HLSL doesn't support fall-though case statements.

Though, now that I think about it, does that count as a fall-though case statement?

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+		"			int4 color = (ras == 0u) ? icolors_0 : icolors_1;\n"
+		"			uint swap = %s << 1;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.rswap).c_str());
+	out.Write(
+		//"			s.RasColor.r = Swap(0, color);\n");
+		"			s.RasColor.r = Swap(%s, color);\n", BitfieldExtract("bpmem_tevksel[swap]", TevKSel().swap1).c_str());
+	out.Write(
+		//"			s.RasColor.g = Swap(1, color);\n");
+		"			s.RasColor.g = Swap(%s, color);\n", BitfieldExtract("bpmem_tevksel[swap]", TevKSel().swap2).c_str());
+	out.Write(
+		"			swap = swap + 1u;\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: I guess this formulation for swap is correct, but could you please use the same way to calculate it for both texture and ras?

File: Source/Core/VideoCommon/UberShaderPixel.cpp (hunk 2 above)
Comment (phire) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: yeah, the rgba to abgr conversion was a quick experiment which lead to shadow link, I'm well aware it's not correct.

Without it things look reasonably normal, except many things switch to black and white

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+
+	// TODO: Indirect Texturing
+	out.Write("\t\t// TODO: Indirect textures\n\n");
+
+	// Disable texturing when there are no texgens (for now)
+	if (numTexgen != 0)
+	{
+	out.Write(
+		"		// Sample texture for stage\n"
+		"		int4 texColor;\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Not used any more

File: Source/Core/VideoCommon/UberShaderPixel.cpp (hunk 2 above)
Comment (phire) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: This is what happens with the correct ordering: 
image

Many things work as expected, but some textures randomly fall back to black and white.

Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: I remember having this white texture with only the first TEV stage, the one with swizzling. So maybe this bug is unrelated?
